rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,329 @@
+   deleted data is collected within a time bound is left for a future exercise.
+2. **Resolving Out-of-Order Operations**: some applications may process incoming requests
+   in a manner that results in out-of-order insertions to SlateDB (stream processing applications
+   are frequently subject to this type of pattern). Using insertion timestamps to resolve which
+   version of a key is more recent may be preferable to using the clock at the time of insertion.
+3. **Historical Reads / MVCC**: while SlateDB provides snapshot capabilities to ensure reads are
+   consistent and repeatable, it requires a snapshot to have been taken at the desired time. 
+   increasing the number of snapshots adds significant storage overhead. Instead, it is possible
+   to trade read-latency for space amplification by maintaining multiple versions of keys at
+   different timestamps to allow for similar "historical" read operations.
+
+## Goals
+
+- Define the API for specifying timestamps via user defined clocks
+- Define the changes to the on-disk format of stored rows
+- Outline mechanism to support Time-To-Live (TTL) via Compaction Filters
+
+## Out of Scope
+
+While the following are out of scope for the initial implementation, the design should
+not preclude:
+
+- Handling out-of-order insertions is not in scope for this RFC, though we make sure the design
+  accounts for the future possibility (see rejected alternatives section)
+- Outlining compaction strategies that are optimized for TTL (such as Cassandra's
+  [TWCS](https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html)). This
+  will require additional metadata per-SST and a compaction strategy to match, but is otherwise
+  possible with what's outlined in the scope of this design.
+- Outlining compaction strategies that target efficient deletes (such as outlined
+  in [Lethe](https://disc-projects.bu.edu/lethe/))
+- Ensuring deletions are resolved within a time bound (a TTL does not ensure the data is deleted
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+We will make the following changes to `DbOptions`:
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+Note that the clock is expected to return monotonically increasing (greater than or equal to
+previous) values. This will be enforced by wrapping the supplied `Clock` implementation that
+tracks the previous values returned.
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
```

Comment:
When using a custom implementation of `Clock`, one of the important case `now()` would be called is when inserting the row. People who use custom implementation would want more control over the specific timestamp used for a write operation, instead of letting the DB get `now()`.

Later on, `write_options` adds a TTL.  Discussion in #218 [talks](https://github.com/slatedb/slatedb/issues/218#issuecomment-2368622047) about the actual timestamp being passed in write option.

Shouldn't there be another parameter to `write_option`?

Reply:
I see this explained in Rejected Alternatives section.

Out of order write described there does make things complex.
Disabling the ability to concretely control the timestamp for row, makes it too strict.

Cases such as reading a Kafka topic, with LogAppend timestamp and intending to use Kafka timestamp as the timestamp for record cannot be done. I don't have a proposal yet.

Reply:
The intention with this design (persisting the timestamp with every row) gives us the future flexibility to easily extend the proposal to row-level insertion timestamps. As far as reading form a Kafka topic, even most stream processors don't handle out of order insertions really elegantly and instead either use watermarks to indicate passage of time or have the concept of "stream time" which advances monotonically based on the most advances timestamp seen to date. I think it's not necessary to cover this in the scope of this proposal, but please let me know if you feel otherwise!

Reply:
I am reviewing the RFC again, and the stated scope looks good to me as is.

Here is one case that I felt would benefit from having more control on the `ts` used for the row.

1. A clock is supplied, its implementation returns the maximum `LogAppendTime` seen so far by the application. The application is reading from a Kafka topic. It will be naturally monotonically non decreasing.
2. A batch of 5 messages are retrieved, they had timestamp `t1..t5`. Clock's now is set to `t5`. These are inserted to the DB from different tasks concurrently.
3. All keys would have a timestamp `t5`, since clock was updated to `t5`. Say TTL was 7 units, and `now()` is `8`. Key with `LogAppendTime = t1` (i.e. first key) must be expired, whereas `t2` and others shouldn't be. 
4. Since all the messages have `t5` none of them would be expired.

Reply:
Applications that need this level of control could use a application level filter on top of TTL, this might anyways be necessary since TTL is not used as a filter. So this might end up as a corner case use case that doesn't need to be handled natively.

Reply:
> could use a application level filter on top of TTL

this is exactly how we do it in Responsive with Cassandra -- db-level TTL is typically a physical consideration more so than a logical one (rocksDB, for example, doesn't give any guarantees around when the cleanup will actually happen). I think we have opportunity to do better but having an application-level logical TTL is a pretty common pattern to ensure correctness in situations where that really matters.